flowmachine/flowmachine/core/cache.py

```diff
@@ -23,6 +34,168 @@
 
 
+def write_query_to_cache(
+    *,
+    name: str,
+    redis: StrictRedis,
+    query: "Query",
+    connection: "Connection",
+    ddl_ops_func: Callable[[str, str], List[str]],
+    write_func: Callable[[List[str], Engine], float],
+    schema: Optional[str] = "cache",
+    sleep_duration: Optional[int] = 1,
+) -> "Query":
+    """
+    Write a Query object into a postgres table and update the cache metadata about it.
+    Attempts to update the query's state in redis to executing, and if successful, tries to run
+    it. If unable to update the state to executing, will block until the query is in a completed,
+    errored, or cancelled state.
+
+    Parameters
+    ----------
+    name : str
+        Name of the table to write to
+    redis : StrictRedis
+        Redis connection to use to update state
+    query : Query
+        Query object to write
+    connection : Connection
+        Flowmachine connection to use for writing
+    ddl_ops_func : Callable[[str, str], List[str]]
+        Function that will be called to generate a list of SQL statements to run. Should accept name and
+        schema as arguments and return a list of SQL strings.
+    write_func : Callable[[List[str], Engine], float]
+        Function which will be called with the result of ddl_ops_func to perform the actual write.
+        Should take a list of SQL strings and an SQLAlchemy Engine as arguments and return the runtime
+        of the query.
+    schema : str, default "cache"
+        Name of the schema to write to
+    sleep_duration : int, default 1
+        Number of seconds to wait between polls when monitoring a query being written from elsewhere
+
+    Returns
+    -------
+    Query
+        The query object which was written once the write has completed successfully.
+
+
+    Raises
+    ------
+    QueryCancelledException
+        If execution of the query was interrupted by a user
+    QueryErroredException
+        If running the ddl ops failed
+    StoreFailedException
+        If something unexpected went wrong while storing the query
+
+    Notes
+    -----
+    This is a _blocking function_, and will not return until the query is no longer in an executing state.
+
+    """
+    logger.debug(f"Trying to switch '{query.md5}' to executing state.")
+    q_state_machine = QueryStateMachine(redis, query.md5)
+    current_state, this_thread_is_owner = q_state_machine.execute()
+    if this_thread_is_owner:
+        logger.debug(f"In charge of executing '{query.md5}'.")
+        query_ddl_ops = ddl_ops_func(name, schema)
+        logger.debug("Made SQL.")
+        con = connection.engine
+        with con.begin():
+            try:
+                plan_time = write_func(query_ddl_ops, con)
+                logger.debug("Executed queries.")
+            except Exception as e:
+                q_state_machine.raise_error()
+                logger.error(f"Error executing SQL. Error was {e}")
+                raise e
+            if schema == "cache":
```

Comment:
Is there a reason to only conditionally write cache metadata? I'm a little confused about the meaning of this `schema` argument. I assumed its purpose was to allow a different name for the cache schema (e.g. for testing), but it looks like it's passed in from `Query.to_sql()` and `ModelResult.to_sql()` and is basically the schema that the queries are stored in? I trust that it does the right thing but I'm not sure I fully follow the logic.